src/Development/IDE/Core/Shake.hs

```diff
@@ -377,57 +386,114 @@ shakeProfile :: IdeState -> FilePath -> IO ()
+shakeEnqueue :: IdeState -> Action a -> IO (IO a)
+shakeEnqueue IdeState{shakeSession} act =
+    withMVar shakeSession $ \s -> runInShakeSession s act
+
+-- Set up a new 'ShakeSession' with a set of initial actions.
+-- Will crash if there is an existing 'ShakeSession' running.
+-- Reports progress on the set of initial actions only.
+newSession :: IdeState -> [Action a] -> IO (ShakeSession, IO [a])
+newSession IdeState{shakeExtras=ShakeExtras{..}, ..} acts = do
+    -- A work queue for actions added via 'runInShakeSession'
+    actionQueue :: TQueue (Action ()) <- newTQueueIO
+
+    start <- offsetTime
+    let
+        -- A daemon-like action used to inject additional work
+        -- Runs actions from the work queue sequentially
+        pumpAction =
+            forever $ join $ liftIO $ atomically $ readTQueue actionQueue
+
+        progressRun
+          | reportProgress = lspShakeProgress ideTesting getLspId eventer inProgress
+          | otherwise = return ()
+
+        workRun restore = withAsync progressRun $ \progressThread -> do
+          let acts' =
+                [ [] <$ pumpAction
+                , parallel acts
+                    -- For the purposes of progress reporting, only 'acts' matters.
+                    -- When done, cancel the progressThread to indicate completion
+                    <* liftIO (cancel progressThread)
+                ]
+          res <- try (restore $ shakeRunDatabaseProfile shakeProfileDir shakeDb acts')
+          runTime <- start
+          let res' = case res of
+                      Left e -> "exception: " <> displayException e
+                      Right _ -> "completed"
+              profile = case res of
+                      Right (_, Just fp) ->
+                          let link = case filePathToUri' $ toNormalizedFilePath' fp of
+                                          NormalizedUri _ x -> x
+                          in ", profile saved at " <> T.unpack link
+                      _ -> ""
+              -- Wrap up in a thread to avoid calling interruptible
+              -- operations inside the masked section
+          let wrapUp = do
+                  logDebug logger $ T.pack $
+                      "Finishing shakeRun (took " ++ showDuration runTime ++ ", " ++ res' ++ profile ++ ")"
+          return (fst <$> res, wrapUp)
+
+    -- Do the work in a background thread
+    workThread <- asyncWithUnmask workRun
+    -- run the wrap up unmasked
+    _ <- async $ do
+        (_, wrapUp) <- wait workThread
+        wrapUp
+
+    -- 'runInShakeSession' is used to append work in this Shake session
+    --  The session stays open until 'cancelShakeSession' is called
+    --  This should only be necessary iff the (virtual) filesystem has changed
+    let runInShakeSession :: forall a . Action a -> IO (IO a)
+        runInShakeSession act = do
+              res <- newEmptyMVar
+              atomically $ writeTQueue actionQueue (act >>= liftIO . putMVar res)
```

Comment:
If `act` throws an exception you will never fill the MVar. Is that intentional? It seems like this could lead to a deadlock.

Reply:
I've pushed a change to propagate the exception